evap/student/tests/test_views.py

```diff
@@ -559,3 +561,136 @@ def test_xmin_of_all_answers_is_updated(self):
+        self.assertTrue(self.evaluation.is_dropout_allowed)
+        self.assertTrue(
+            self.evaluation.general_contribution.questionnaires.filter(type=Questionnaire.Type.DROPOUT).exists()
+        )
+
+        normal_questionnaires = self.evaluation.general_contribution.questionnaires.exclude(
+            type=Questionnaire.Type.DROPOUT
+        ).all()
+        self.evaluation.general_contribution.questionnaires.set(normal_questionnaires)
+
+        self.assertFalse(self.evaluation.is_dropout_allowed)
+        self.assertFalse(
+            self.evaluation.general_contribution.questionnaires.filter(type=Questionnaire.Type.DROPOUT).exists()
+        )
+
+    def test_dropout_forbidden(self):
+        normal_questionnaires = self.evaluation.general_contribution.questionnaires.exclude(
+            type=Questionnaire.Type.DROPOUT
+        ).all()
+        self.evaluation.general_contribution.questionnaires.set(normal_questionnaires)
+
+        self.app.get(url=reverse("student:drop", args=[self.evaluation.id]), user=self.user, status=400)
+
+    def test_dropping_out_increments_dropout_counter(self):
+        self.assertEqual(self.evaluation.dropout_count, 0, "dropout count should be initially zero")
+
+        form = self.app.get(url=reverse("student:drop", args=[self.evaluation.id]), user=self.user, status=200).forms[
+            "student-vote-form"
+        ]
+        field_id = answer_field_id(
+            self.evaluation.general_contribution, self.dropout_questionnaire, self.dropout_question
+        )
+        form[field_id] = NO_ANSWER  # dropout question must be answered
+        form.submit()
+        self.evaluation = Evaluation.objects.get(pk=self.evaluation.pk)
+
+        self.assertEqual(self.evaluation.dropout_count, 1, "dropout count should increment with dropout")
+
+        form = self.app.get(url=reverse("student:vote", args=[self.evaluation.id]), user=self.user2, status=200).forms[
+            "student-vote-form"
+        ]
+        form.submit()
+        self.evaluation = Evaluation.objects.get(pk=self.evaluation.pk)
+
+        self.assertEqual(self.evaluation.dropout_count, 1, "dropout count should not change on normal vote")
+
+    def test_choice_fields_correctly_populated_by_server(self):
+        form = self.app.get(url=reverse("student:drop", args=[self.evaluation.id]), user=self.user, status=200).forms[
+            "student-vote-form"
+        ]
+
+        normal_question_id = answer_field_id(
+            self.evaluation.general_contribution, self.normal_questionnaire, self.normal_question
+        )
+        dropout_question_id = answer_field_id(
+            self.evaluation.general_contribution, self.dropout_questionnaire, self.dropout_question
+        )
+
+        self.assertEqual(form[normal_question_id].value, str(NO_ANSWER))
+        self.assertIsNone(form[dropout_question_id].value)
+
+        form[normal_question_id] = -1
+
+        form = form.submit().forms["student-vote-form"]
```

Comment:
just to be sure, this form is returned with errors so we should have the following, right?
```suggestion
        form = form.submit().forms["student-vote-form"]
        self.assertEqual(self.evaluation.dropout_count, 0)
```